Step1 2단계

tmddnrdl333 opened the pull request:
step1 1단계 피드백 반영 및 2단계 구현했습니다.

tmddnrdl333 committed:
feat: step1 1단계 코틀린 기초

- Person 클래스 생성
- Kotest StringSpec으로 테스트

tmddnrdl333 committed:
refactor: step1 1단계 코틀린 기초

- Person 클래스 name not null로 변경
- Person 클래스 nickname private으로 변경
- 테스트 보완

tmddnrdl333 committed:
refactor: step1 1단계 코틀린 기초

- Person 클래스 name not null로 변경
- Person 클래스 nickname private으로 변경
- 테스트 보완

jsw0824 committed:
feat: step1 2단계 문자열 계산기

- 문자열을 받아오고 계산을 수행하는 Calculator 클래스 생성
- 각 기능에 대한 테스트 구현

jsw0824 committed:
Merge branch 'step2' of https://github.com/tmddnrdl333/kotlin-racingcar into step1

lhs0610 commented on src/main/kotlin/Calculator.kt:
```diff
@@ -0,0 +1,59 @@
+class Calculator(str: String?) {
+
+    private var num: Int = 0
+    private var op: Char = '+'
+    private val list: List<String>
```
이렇게 변수들을 선언한 이유가 있을까요?

tmddnrdl333 replied:
자세한 피드백 주셔서 감사합니다!
읽으면서 설계에 구멍이 많은 것 같아서 재설계해서 구현해봤습니다.
감사합니다!

lhs0610 commented on src/main/kotlin/Calculator.kt:
```diff
@@ -0,0 +1,59 @@
+class Calculator(str: String?) {
```
계산기가 하나의 문자열마다 새로 생성될 필요가 있을까요? 메서드에서만 문자열을 받아도 될 것 같네요!

lhs0610 commented on src/main/kotlin/Calculator.kt:
```diff
@@ -0,0 +1,59 @@
+        return num
+    }
+
+    private fun operate(op: Char?, num2: Int?) {
```
op와 num2라는 이름은 좀 직관적이지 않은 것 같네요. 좀 더 이름을 명확하게 지을 수 없을까요?
그리고 nullable 타입으로 선언한 이유가 있을까요? 별도의 사유가 있는게 아니라면 non-null 타입으로 하는 걸 권장드립니다.

lhs0610 commented on src/main/kotlin/Calculator.kt:
```diff
@@ -0,0 +1,59 @@
+    }
+
+    private fun operate(op: Char?, num2: Int?) {
+        op ?: throwIfNoOperator()
+        num2 ?: throwIfNoNumber()
```
require문을 이용해 예외 체킹을 할 수 있어요([link](https://catsbi.oopy.io/2885f744-600e-49ea-a6a8-e321a31df527#05d7fa96-5738-43d1-9d9d-d5e5e4f6ce41))

tmddnrdl333 replied:
감사합니다! 다음 커밋에 적용해보았습니다.

lhs0610 commented on src/main/kotlin/Calculator.kt:
```diff
@@ -0,0 +1,59 @@
+        op ?: throwIfNoOperator()
+        num2 ?: throwIfNoNumber()
+        when (op) {
+            '+' -> num += num2!!
+            '-' -> num -= num2!!
+            '*' -> num *= num2!!
+            '/' -> num /= num2!!
```
각 문자열에 대한 연산 책임을 when 절에서 하드코딩으로 직접 하게 될 경우 유지보수성 측면에서 떨어질 수 있어요. 
만약 여기서 제곱이라거나 다른 연산이 추가된다면 이 코드를 수정해야하는데 이 경우, OCP원칙을 위반할 수 있겠죠?
열거타입으로 분리를 해보면 어떨까요

tmddnrdl333 replied:
다음 커밋에 열거형에서 인터페이스를 구현하는 형식으로 작성해봤습니다!

lhs0610 commented on src/main/kotlin/Calculator.kt:
```diff
@@ -0,0 +1,59 @@
+        }
+    }
+
+    private fun setOp(el: String) {
+        when (el) {
+            "+", "-", "*", "/" -> op = el[0]
+            else -> throwIfNoOperator()
+        }
+    }
```
여기도 위와같이 사칙연산에 대한 하드코등 문제, 인스턴스변수 공유문제등이 있을 것 같네요.

lhs0610 commented on src/main/kotlin/Calculator.kt:
```diff
@@ -0,0 +1,59 @@
+        return list
+    }
+
+    private fun throwIfNotExpression() {
+        throw IllegalArgumentException("Expression Required")
+    }
+
+    private fun throwIfNoNumber() {
+        throw IllegalArgumentException("Number Required")
+    }
+
+    private fun throwIfNoOperator() {
+        throw IllegalArgumentException("Operator Required")
+    }
```
예외를 던지는 것 뿐인 함수는 어떤 의미가 있을까요?

lhs0610 commented on src/test/kotlin/Test.kt:
```diff
@@ -0,0 +1,61 @@
+import io.kotest.core.spec.style.StringSpec
+import io.kotest.matchers.shouldBe
+
+class Test : StringSpec({
```
0으로 나누거나 0을 나누려는 경우, 공백으로 문자열 분리가 안되있는 경우 연산자가 연달아 있거나 피연산자가 연달아 있는 경우, 맨 앞이나 맨 끝이 피연산자가 아니라 연산자인경우 등에 대한 테스트도 있어야 할 듯 싶네요.

tmddnrdl333 replied:
다음 커밋에 말씀 주신 엣지 케이스들에 대한 테스트도 작성해봤습니다!

lhs0610 reviewed (changes requested):
승욱님 2단계 미션 잘 구현해주셨는데, 
- 하나의 클래스에 모든 책임이 다 있어 SRP 원칙에 위배된다. 
- 매직 리터럴이 그대로 하드코딩 되어 유지보수가 힘들다. 

그리고 추가적인 기타 피드백을 남겨드렸으니 확인 후 다시 리뷰요청 부탁드려요.

jsw0824 committed:
refactor: step1 2단계 문자열 계산기

- 기존 Calculator 클래스 대신 각 수식을 인스턴스로 생성하는 MathExpression.kt 작성
- 기존 Calculator 클래스에서의 계산을 클래스 대신 Top-level function 형태로 구현
- 연산자와 그 연산 방식을 관리하는 Operator.kt 열거형으로 작성
- require 문 적용
- 여러 엣지 케이스들에 대한 테스트 코드 작성

lhs0610 commented on src/main/kotlin/component/MathExpression.kt:
```diff
@@ -0,0 +1,45 @@
+
+const val INVALID_EXPRESSION = "Invalid Expression"
+
+class MathExpression(str: String?) {
```
- 변수명은 의미있게 지어주세요.  
- 초기화블럭에서 null check해서 예외를 던질꺼면 그냥 여기서 non-null 타입으로 선언하면 되지 않을까요?

tmddnrdl333 replied:
실제로 문자열 계산기를 만든다면 그게 맞을 것 같습니다.
근데 제시문에서 입력 값이 null인 경우도 상정하고 있는 것 같아서 그렇게 구현한 것 같습니다..

다음 커밋에는 참고하여 non-null 로 변경 적용하겠습니다!

lhs0610 commented on src/main/kotlin/component/MathExpression.kt:
```diff
@@ -0,0 +1,45 @@
+
+    init {
+        str ?: throw IllegalArgumentException(INVALID_EXPRESSION)
+        val list = str.split(" ")
```
여기는 또 하드코딩을 하셨네요 🤔

lhs0610 commented on src/main/kotlin/component/MathExpression.kt:
```diff
@@ -0,0 +1,45 @@
+        str ?: throw IllegalArgumentException(INVALID_EXPRESSION)
+        val list = str.split(" ")
+
+        for ((index, el) in list.withIndex()) {
+            if (index % 2 == 0) {
+                numList.add(el.toValidatedNum())
+            } else {
+                operatorList.add(el.toValidatedOperator())
+            }
+        }
```
- 반복문안에 분기문이 있어요. 그러다보니 indent가 3이 되는데 어떻게 줄일 수 없을까요? 
- 한 로직에 연산자 추출과 피연산자 추출 로직 두 개를 모두 책임지고 있어요. 분리해보면 어떨까요?  

다음 글을 참고해서 지키도록 해보시죠 ([link](https://catsbi.oopy.io/bf003ff6-2912-4714-8ac2-44eeb7becc93))

lhs0610 commented on src/main/kotlin/component/MathExpression.kt:
```diff
@@ -0,0 +1,45 @@
+    }
+
+    private fun String.toValidatedOperator(): Operator {
+        require(this.length == 1) {
+            INVALID_EXPRESSION
+        }
```
이런 검증역할도 Operator가 알아서 해도 되지 않을까요?

lhs0610 commented on src/main/kotlin/component/MathExpression.kt:
```diff
@@ -0,0 +1,45 @@
+    }
+
+    private fun validateExpression(): Boolean {
+        require(operatorList.size > 0 && operatorList.size + 1 == numList.size) {
```
이 메서드를 호출하는 부분이 이미 require인데, 여기서도 require를 쓸 필요가 있을까요?

lhs0610 commented on src/main/kotlin/enums/Operator.kt:
```diff
@@ -0,0 +1,36 @@
+
+const val INVALID_EXPRESSION = "Invalid Expression"
+
+enum class Operator(val operator: String) : BinaryOperator<Int> {
```
람다 표현식을 통해 더 간결하게 코드 작성이 가능할 것 같아요.

lhs0610 commented on src/main/kotlin/enums/Operator.kt:
```diff
@@ -0,0 +1,36 @@
+    },
+    DIVIDE("/") {
+        override fun apply(t: Int, u: Int): Int {
+            return t / u
```
0을 나누거나 할 때 ArithmeticException이 발생할 것 같네요

lhs0610 commented on src/main/kotlin/enums/Operator.kt:
```diff
@@ -0,0 +1,36 @@
+    };
+
+    companion object {
+        fun of(operator: String): Operator {
+            return values().firstOrNull {
+                it.operator == operator
+            } ?: throw IllegalArgumentException(INVALID_EXPRESSION)
+        }
```
원하는 타입이 없을 때 예외를 던지는 것 보다는 NONE과 같이 없다는 결과의 Enum을 만들어 제공하는게 더 나은 선택일 수 있어요

lhs0610 reviewed (approved):
승욱님, 추가적인 패드백 남겨드렸으니 확인 후 다음 미션 진행하시며 같이 반영해보도록 하시죠. 
이만 merge하겠습니다.
